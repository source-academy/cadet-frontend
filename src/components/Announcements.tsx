/**
 * This component is on hold until an endpoint is implemented at
 * source-academy/cadet which is able to download and relay news and material
 * from lumiNUS to us.
 */
import { Card, NonIdealState, Spinner, Text } from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import * as React from 'react';
import ContentDisplay, { IContentDisplayProps } from './commons/ContentDisplay';

export type Announcement = {
  author: string;
  title: string;
  content: string;
  pinned: boolean;
};

export interface IAnnouncementsProps {
  announcements?: Announcement[];
  handleAnnouncementsFetch: () => void;
}

interface IAnnouncementCardProps {
  announcements?: Announcement[];
}

class Announcements extends React.Component<IAnnouncementsProps, {}> {
  public render() {
    const props: IContentDisplayProps = {
      display: <AnnouncementCard announcements={this.props.announcements} />,
      loadContentDispatch: this.props.handleAnnouncementsFetch
    };
    return <ContentDisplay {...props} />;
  }
}

export const AnnouncementCard: React.SFC<IAnnouncementCardProps> = props => {
  if (props.announcements === undefined) {
    return <NonIdealState description="Fetching announcements..." visual={<Spinner />} />;
  } else if (props.announcements.length === 0) {
    return <NonIdealState title="There are no announcements." visual={IconNames.FEED} />;
  } else {
    const cards = props.announcements.map((ann, index) => (
      <div key={index}>
        <Card>
          <h4>{ann.title}</h4>
          <Text>{ann.content}</Text>
        </Card>
      </div>
    ));
    return <>{cards}</>;
  }
};

export default Announcements;
