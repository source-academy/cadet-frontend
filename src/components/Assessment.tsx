import { Button, Dialog, NonIdealState, Spinner } from '@blueprintjs/core'
import { IconNames } from '@blueprintjs/icons'
import * as React from 'react'

import Workspace from '../containers/workspace'
import { SideContentTab } from './workspace/side-content'

export type AssessmentInfo = {
  longSummary: string
  dueDate: string
  studentBriefed: boolean
}

export type StateProps = {
  missionId: number
  assessmentInfo?: AssessmentInfo
}

export type DispatchProps = {
  handleAssessmentInfoFetch: (missionId: number) => void
}

export type AssessmentProps = StateProps & DispatchProps

class Assessment extends React.Component<AssessmentProps, { showOverlay: boolean }> {
  public state = { showOverlay: true }

  public render() {
    if (this.props.assessmentInfo === undefined) {
      this.props.handleAssessmentInfoFetch(this.props.missionId)
      return <NonIdealState description="Getting mission ready..." visual={<Spinner />} />
    }
    const briefing = <> ({this.props.assessmentInfo.longSummary}) </>
    const overlay = (
      <Dialog
        className="mission-briefing"
        isOpen={this.state.showOverlay && !this.props.assessmentInfo.studentBriefed}
      >
        <>
          {briefing}
          <Button
            // tslint:disable-next-line jsx-no-lambda
            onClick={() => this.setState({ showOverlay: false })}
            text="Primary"
          />
        </>
      </Dialog>
    )
    const briefingTab: SideContentTab = {
      label: 'Briefing',
      icon: IconNames.BRIEFCASE,
      body: briefing
    }
    return (
      <>
        {overlay}
        <Workspace sideContentTabs={[briefingTab]} />
      </>
    )
  }
}

export default Assessment
