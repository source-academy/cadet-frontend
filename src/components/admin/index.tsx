import { Card } from '@blueprintjs/core'
import * as React from 'react'
import { Redirect, Route, Switch } from 'react-router'

import AnnouncementManager from '../../containers/admin/AnnouncementManagerContainer'
import LoginContainer from '../../containers/LoginContainer'
import AdminNavigationBar from './NavigationBar'

const redirectToAnnouncements = () => <Redirect to="/admin/announcements" />

const Admin: React.SFC<{}> = () => {
  const redirectTo404 = () => <Redirect to="/404" />

  return (
    <div className="Admin">
      <LoginContainer />
      <AdminNavigationBar />
      <div className="row center-xs">
        <div className="col-xs-12 admin-content-parent">
          <Card className="admin-content" elevation={1}>
            <Switch>
              <Route path="/admin/announcements" component={AnnouncementManager} />
              <Route exact={true} path="/admin" component={redirectToAnnouncements} />
              <Route component={redirectTo404} />
            </Switch>
          </Card>
        </div>
      </div>
    </div>
  )
}

export default Admin
