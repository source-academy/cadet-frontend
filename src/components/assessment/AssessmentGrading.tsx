import { Text } from '@blueprintjs/core'
import * as React from 'react'
import { getPrettyDate } from '../../utils/dateHelpers'
import Markdown from '../commons/Markdown'

type AssessmentGradingProps = OwnProps

export type OwnProps = {
  comment: string | null
  graderName: string
  gradedAt: string
  xp: number
  grade: number
  maxGrade: number
  maxXp: number
}

class AssessmentGrading extends React.Component<AssessmentGradingProps, {}> {
  constructor(props: AssessmentGradingProps) {
    super(props)
  }

  public render() {
    return (
      <div className="AssessmentGrading">
        <div className="assessment-grading-table">
          <table>
            <tbody>
              <tr>
                <th>Grade:</th>
                <td>
                  <Text>
                    {this.props.grade} / {this.props.maxGrade}
                  </Text>
                </td>
              </tr>

              <tr>
                <th>XP:</th>
                <td>
                  <Text>
                    {this.props.xp} / {this.props.maxXp}
                  </Text>
                </td>
              </tr>
            </tbody>
          </table>

          {this.props.comment !== null ? (
            <div>
              <br />
              <th>Comment:</th>
              <p>
                <pre>
                  <Markdown content={this.props.comment} />
                </pre>
              </p>
            </div>
          ) : null}

          <br />

          <div className="assessment-grading-info">
            <Text>
              Graded by <b>{this.props.graderName}</b> on {getPrettyDate(this.props.gradedAt)}
            </Text>
          </div>
        </div>
      </div>
    )
  }
}

export default AssessmentGrading
