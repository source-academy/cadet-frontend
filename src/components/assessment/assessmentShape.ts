/*
 * Used to display information regarding an assessment in the UI.
 *
 * @property closeAt an ISO 8601 compliant date string specifiying when
 *   the assessment closes
 * @property openAt an ISO 8601 compliant date string specifiying when
 *   the assessment opens
 */
export interface IAssessmentOverview {
  category: AssessmentCategory;
  closeAt: string;
  coverImage: string;
  grade: number;
  id: number;
  maxGrade: number;
  maxXp: number;
  openAt: string;
  title: string;
  shortSummary: string;
  status: AssessmentStatus;
  story: string | null;
  xp: number;
  gradingStatus: GradingStatus;
}

export enum AssessmentStatuses {
  not_attempted = 'not_attempted',
  attempting = 'attempting',
  attempted = 'attempted',
  submitted = 'submitted'
}

export type AssessmentStatus = keyof typeof AssessmentStatuses;

export enum GradingStatuses {
  none = 'none',
  grading = 'grading',
  graded = 'graded'
}

export type GradingStatus = keyof typeof GradingStatuses;

/*
 * Used when an assessment is being actively attempted/graded.
 */
export interface IAssessment {
  category: AssessmentCategory;
  id: number;
  longSummary: string;
  missionPDF: string;
  questions: IQuestion[];
  title: string;
}

/* The different kinds of Assessments available */
export enum AssessmentCategories {
  Contest = 'Contest',
  Mission = 'Mission',
  Path = 'Path',
  Sidequest = 'Sidequest'
}

export type AssessmentCategory = keyof typeof AssessmentCategories;

export interface IProgrammingQuestion extends IQuestion {
  answer: string | null;
  solutionTemplate: string;
  type: 'programming';
}

export interface IMCQQuestion extends IQuestion {
  solution: number | null;
  answer: number | null;
  choices: MCQChoice[];
  type: 'mcq';
}

export interface IQuestion {
  answer: string | number | null;
  comment: string | null;
  content: string;
  id: number;
  library: Library;
  type: QuestionType;
  grader: {
    name: string;
    id: number;
  };
  gradedAt: string;
  xp: number;
  grade: number;
  maxGrade: number;
  maxXp: number;
}

export type MCQChoice = {
  content: string;
  hint: string | null;
};

/* The two kinds of Questions available */
export enum QuestionTypes {
  programming = 'programming',
  mcq = 'mcq'
}
export type QuestionType = keyof typeof QuestionTypes;

/** Constants for external library names */
export enum ExternalLibraryNames {
  NONE = 'NONE',
  TWO_DIM_RUNES = 'TWO_DIM_RUNES',
  THREE_DIM_RUNES = 'THREE_DIM_RUNES',
  CURVES = 'CURVES',
  SOUND = 'SOUND',
  STREAMS = 'STREAMS'
}

export type ExternalLibraryName = keyof typeof ExternalLibraryNames;

type ExternalLibrary = {
  name: ExternalLibraryName;
  symbols: string[];
};

export type Library = {
  chapter: number;
  external: ExternalLibrary;
  globals: Array<[string, any]>;
};
