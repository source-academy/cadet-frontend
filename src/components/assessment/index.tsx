import {
  Button,
  ButtonGroup,
  Card,
  Classes,
  Collapse,
  Dialog,
  Elevation,
  Icon,
  IconName,
  Intent,
  NonIdealState,
  Position,
  Spinner,
  Text,
  Tooltip
} from '@blueprintjs/core';
import { IconNames } from '@blueprintjs/icons';
import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { NavLink } from 'react-router-dom';

import defaultCoverImage from '../../assets/default_cover_image.jpg';
import AssessmentWorkspaceContainer from '../../containers/assessment/AssessmentWorkspaceContainer';
import { beforeNow, getPrettyDate } from '../../utils/dateHelpers';
import { assessmentCategoryLink, stringParamToInt } from '../../utils/paramParseHelpers';
import {
  AssessmentCategory,
  AssessmentStatuses,
  GradingStatuses,
  IAssessmentOverview
} from '../assessment/assessmentShape';
import { OwnProps as AssessmentProps } from '../assessment/AssessmentWorkspace';
import { controlButton } from '../commons';
import ContentDisplay from '../commons/ContentDisplay';
import Markdown from '../commons/Markdown';

const DEFAULT_QUESTION_ID: number = 0;

export interface IAssessmentWorkspaceParams {
  assessmentId?: string;
  questionId?: string;
}

export interface IAssessmentProps
  extends IDispatchProps,
    IOwnProps,
    RouteComponentProps<IAssessmentWorkspaceParams>,
    IStateProps {}

export interface IDispatchProps {
  handleAssessmentOverviewFetch: () => void;
  handleSubmitAssessment: (id: number) => void;
}

export interface IOwnProps {
  assessmentCategory: AssessmentCategory;
}

export interface IStateProps {
  assessmentOverviews?: IAssessmentOverview[];
  isStudent: boolean;
}

type State = {
  betchaAssessment: IAssessmentOverview | null;
  showClosedAssessments: boolean;
  showOpenedAssessments: boolean;
  showUpcomingAssessments: boolean;
};

class Assessment extends React.Component<IAssessmentProps, State> {
  /**
   * Initialize state
   */
  public constructor(props: IAssessmentProps) {
    super(props);
    this.state = {
      betchaAssessment: null,
      showClosedAssessments: false,
      showOpenedAssessments: true,
      showUpcomingAssessments: true
    };
  }

  public render() {
    const assessmentId: number | null = stringParamToInt(this.props.match.params.assessmentId);
    const questionId: number =
      stringParamToInt(this.props.match.params.questionId) || DEFAULT_QUESTION_ID;

    // If there is an assessment to render, create a workspace. The assessment
    // overviews must still be loaded for this, to send the due date.
    if (assessmentId !== null && this.props.assessmentOverviews !== undefined) {
      const overview = this.props.assessmentOverviews.filter(a => a.id === assessmentId)[0];
      const assessmentProps: AssessmentProps = {
        assessmentId,
        questionId,
        notAttempted: overview.status === AssessmentStatuses.not_attempted,
        closeDate: overview.closeAt
      };
      return <AssessmentWorkspaceContainer {...assessmentProps} />;
    }

    // Otherwise, render a list of assessments to the user.
    let display: JSX.Element;
    if (this.props.assessmentOverviews === undefined) {
      display = <NonIdealState description="Fetching assessment..." visual={<Spinner />} />;
    } else if (this.props.assessmentOverviews.length === 0) {
      display = <NonIdealState title="There are no assessments." visual={IconNames.FLAME} />;
    } else {
      /** Upcoming assessments, that are not released yet. */
      const isOverviewUpcoming = (overview: IAssessmentOverview) =>
        !beforeNow(overview.closeAt) && !beforeNow(overview.openAt);
      const upcomingCards = this.props.assessmentOverviews
        .filter(isOverviewUpcoming)
        .map((overview, index) =>
          makeOverviewCard(overview, index, this.setBetchaAssessment, !this.props.isStudent, false)
        );

      /** Opened assessments, that are released and can be attempted. */
      const isOverviewOpened = (overview: IAssessmentOverview) =>
        !beforeNow(overview.closeAt) &&
        beforeNow(overview.openAt) &&
        overview.status !== AssessmentStatuses.submitted;
      const openedCards = this.props.assessmentOverviews
        .filter(overview => isOverviewOpened(overview))
        .map((overview, index) =>
          makeOverviewCard(overview, index, this.setBetchaAssessment, true, false)
        );

      /** Closed assessments, that are past the due date or cannot be attempted further. */
      const closedCards = this.props.assessmentOverviews
        .filter(overview => !isOverviewOpened(overview) && !isOverviewUpcoming(overview))
        .map((overview, index) =>
          makeOverviewCard(overview, index, this.setBetchaAssessment, true, true)
        );

      /** Render cards */
      const upcomingCardsCollapsible =
        upcomingCards.length > 0 ? (
          <>
            {collapseButton(
              'Upcoming',
              this.state.showUpcomingAssessments,
              this.toggleUpcomingAssessments
            )}
            <Collapse isOpen={this.state.showUpcomingAssessments}>{upcomingCards}</Collapse>
          </>
        ) : null;
      const openedCardsCollapsible =
        openedCards.length > 0 ? (
          <>
            {collapseButton('Open', this.state.showOpenedAssessments, this.toggleOpenAssessments)}
            <Collapse isOpen={this.state.showOpenedAssessments}>{openedCards}</Collapse>
          </>
        ) : null;
      const closedCardsCollapsible =
        closedCards.length > 0 ? (
          <>
            {collapseButton(
              'Closed',
              this.state.showClosedAssessments,
              this.toggleClosedAssessments
            )}
            <Collapse isOpen={this.state.showClosedAssessments}>{closedCards}</Collapse>
          </>
        ) : null;
      display = (
        <>
          {upcomingCardsCollapsible}
          {openedCardsCollapsible}
          {closedCardsCollapsible}
        </>
      );
    }

    // Define the betcha dialog (in each card's menu)
    const betchaText = this.state.betchaAssessment ? (
      <>
        <p>
          You are about to finalise your submission for the{' '}
          {this.state.betchaAssessment.category.toLowerCase()}{' '}
          <i>
            &quot;
            {this.state.betchaAssessment.title}
            &quot;
          </i>
          .
        </p>
        <p>
          Early submissions grant you additional XP, but{' '}
          <span className="warning">this action is irreversible.</span>
        </p>
      </>
    ) : (
      <>
        <p>You are about to finalise your submission.</p>
        <p>
          Early submissions grant you additional XP, but{' '}
          <span className="warning">this action is irreversible.</span>
        </p>
      </>
    );
    const betchaDialog = (
      <Dialog
        className="betcha-dialog"
        icon={IconNames.ERROR}
        isCloseButtonShown={false}
        isOpen={this.state.betchaAssessment !== null}
        title="Betcha: Early Submission"
      >
        <div className={Classes.DIALOG_BODY}>
          <Text>{betchaText}</Text>
        </div>
        <div className={Classes.DIALOG_FOOTER}>
          <ButtonGroup>
            {controlButton('Cancel', null, this.setBetchaAssessmentNull, { minimal: false })}
            {controlButton('Finalise Submission', null, this.submitAssessment, {
              minimal: false,
              intent: Intent.DANGER
            })}
          </ButtonGroup>
        </div>
      </Dialog>
    );

    // Finally, render the ContentDisplay.
    return (
      <div className="Assessment">
        <ContentDisplay
          display={display}
          loadContentDispatch={this.props.handleAssessmentOverviewFetch}
        />
        {betchaDialog}
      </div>
    );
  }

  private toggleClosedAssessments = () =>
    this.setState({
      ...this.state,
      showClosedAssessments: !this.state.showClosedAssessments
    });

  private toggleOpenAssessments = () =>
    this.setState({
      ...this.state,
      showOpenedAssessments: !this.state.showOpenedAssessments
    });

  private toggleUpcomingAssessments = () =>
    this.setState({
      ...this.state,
      showUpcomingAssessments: !this.state.showUpcomingAssessments
    });

  private setBetchaAssessment = (assessment: IAssessmentOverview | null) =>
    this.setState({
      ...this.state,
      betchaAssessment: assessment
    });

  private setBetchaAssessmentNull = () => this.setBetchaAssessment(null);

  private submitAssessment = () => {
    if (this.state.betchaAssessment) {
      this.props.handleSubmitAssessment(this.state.betchaAssessment.id);
      this.setBetchaAssessmentNull();
    }
  };
}

/**
 * Create a series of cards to display IAssessmentOverviews.
 * @param {IAssessmentOverview} overview the assessment overview to display
 * @param {number} index a unique number for this card (required for sequential rendering).
 *   See {@link https://reactjs.org/docs/lists-and-keys.html#keys}
 * @param setBetchaAssessment a function that handles the side-effect of setting which assessment
 *   is to be set for final submission ("betcha" functionality)
 * @param renderAttemptButton will only render the attempt button if true, regardless
 *   of attempt status.
 */
const makeOverviewCard = (
  overview: IAssessmentOverview,
  index: number,
  setBetchaAssessment: (assessment: IAssessmentOverview | null) => void,
  renderAttemptButton: boolean,
  renderGradingStatus: boolean
) => (
  <div key={index}>
    <Card className="row listing" elevation={Elevation.ONE}>
      <div className="col-xs-3 listing-picture">
        <img
          className={`cover-image-${overview.status}`}
          src={overview.coverImage ? overview.coverImage : defaultCoverImage}
        />
      </div>
      <div className="col-xs-9 listing-text">
        {makeOverviewCardTitle(overview, index, setBetchaAssessment, renderGradingStatus)}
        <div className="row listing-grade">
          <h6>
            {' '}
            {beforeNow(overview.openAt)
              ? `Grade: ${overview.grade} / ${overview.maxGrade}`
              : `Max Grade: ${overview.maxGrade}`}{' '}
          </h6>
        </div>
        <div className="row listing-xp">
          <h6>
            {' '}
            {beforeNow(overview.openAt)
              ? `XP: ${overview.xp} / ${overview.maxXp}`
              : `Max XP: ${overview.maxXp}`}{' '}
          </h6>
        </div>
        <div className="row listing-description">
          <Markdown content={overview.shortSummary} />
        </div>
        <div className="listing-controls">
          <Text className="listing-due-date">
            <Icon className="listing-due-icon" iconSize={12} icon={IconNames.TIME} />
            {beforeNow(overview.openAt)
              ? `Due: ${getPrettyDate(overview.closeAt)}`
              : `Opens at: ${getPrettyDate(overview.openAt)}`}
          </Text>
          {renderAttemptButton ? makeOverviewCardButton(overview) : null}
        </div>
      </div>
    </Card>
  </div>
);

const makeOverviewCardTitle = (
  overview: IAssessmentOverview,
  index: number,
  setBetchaAssessment: (assessment: IAssessmentOverview | null) => void,
  renderGradingStatus: boolean
) => (
  <div className="row listing-title">
    <Text ellipsize={true} className={'col-xs-10'}>
      <h4>
        {overview.title} {renderGradingStatus ? makeGradingStatus(overview.gradingStatus) : null}
      </h4>
    </Text>
    <div className="col-xs-2">{makeSubmissionButton(overview, index, setBetchaAssessment)}</div>
  </div>
);

const makeGradingStatus = (gradingStatus: string) => {
  let iconName: IconName;
  let intent: Intent;
  let tooltip: string;

  switch (gradingStatus) {
    case GradingStatuses.graded:
      iconName = IconNames.TICK;
      intent = Intent.SUCCESS;
      tooltip = 'Fully graded';
      break;

    case GradingStatuses.grading:
      iconName = IconNames.TIME;
      intent = Intent.WARNING;
      tooltip = 'Grading in progress';
      break;

    default:
      iconName = IconNames.CROSS;
      intent = Intent.DANGER;
      tooltip = 'Not graded yet';
      break;
  }

  return (
    <Tooltip content={tooltip} position={Position.RIGHT}>
      <Icon icon={iconName} intent={intent} />
    </Tooltip>
  );
};

const makeSubmissionButton = (
  overview: IAssessmentOverview,
  index: number,
  setBetchaAssessment: (assessment: IAssessmentOverview | null) => void
) => (
  <Button
    disabled={overview.status !== AssessmentStatuses.attempted}
    icon={IconNames.CONFIRM}
    intent={overview.status === AssessmentStatuses.attempted ? Intent.DANGER : Intent.NONE}
    minimal={true}
    // intentional: each menu renders own version of onClick
    // tslint:disable-next-line:jsx-no-lambda
    onClick={() => setBetchaAssessment(overview)}
  >
    Finalize Submission
  </Button>
);

const makeOverviewCardButton = (overview: IAssessmentOverview) => {
  let icon: IconName;
  let label: string;
  switch (overview.status) {
    case AssessmentStatuses.not_attempted:
      icon = IconNames.STEP_FORWARD;
      label = overview.story ? 'Skip Story & Attempt' : 'Attempt';
      break;
    case AssessmentStatuses.attempting:
      icon = IconNames.PLAY;
      label = 'Continue Attempt';
      break;
    case AssessmentStatuses.attempted:
      icon = IconNames.EDIT;
      label = 'Review Attempt';
      break;
    case AssessmentStatuses.submitted:
      icon = IconNames.EYE_OPEN;
      label = 'Review Submission';
      break;
    default:
      // If we reach this case, backend data did not fit IAssessmentOverview
      icon = IconNames.PLAY;
      label = 'Review';
      break;
  }
  return (
    <NavLink
      to={`/academy/${assessmentCategoryLink(
        overview.category
      )}/${overview.id.toString()}/${DEFAULT_QUESTION_ID}`}
    >
      {controlButton(label, icon)}
    </NavLink>
  );
};

const collapseButton = (label: string, isOpen: boolean, toggleFunc: () => void) =>
  controlButton(label, isOpen ? IconNames.CARET_DOWN : IconNames.CARET_RIGHT, toggleFunc, {
    minimal: true,
    className: 'collapse-button'
  });

export default Assessment;
