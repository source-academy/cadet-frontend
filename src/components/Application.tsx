import { decompressFromEncodedURIComponent } from 'lz-string';
import * as qs from 'query-string';
import * as React from 'react';
import { Redirect, Route, RouteComponentProps, Switch } from 'react-router';

import Academy from '../containers/academy';
import Login from '../containers/LoginContainer';
import Playground from '../containers/PlaygroundContainer';
import { Role, sourceChapters } from '../reducers/states';
import { ExternalLibraryName, ExternalLibraryNames } from './assessment/assessmentShape';
import NavigationBar from './NavigationBar';
import NotFound from './NotFound';

export interface IApplicationProps extends IDispatchProps, IStateProps, RouteComponentProps<{}> {}

export interface IStateProps {
  accessToken?: string;
  currentPlaygroundChapter: number;
  role?: Role;
  title: string;
  name?: string;
  currentPlaygroundExternalLibrary: ExternalLibraryName;
}

export interface IDispatchProps {
  handleClearContext: (chapter: number, externalLibraryName: ExternalLibraryName) => void;
  handleEditorValueChange: (val: string) => void;
  handleEnsureLibrariesLoaded: () => void;
  handleLogOut: () => void;
  handlePlaygroundExternalSelect: (external: ExternalLibraryName) => void;
}

class Application extends React.Component<IApplicationProps, {}> {
  public componentDidMount() {
    parsePlayground(this.props);
  }

  public render() {
    return (
      <div className="Application">
        <NavigationBar
          handleLogOut={this.props.handleLogOut}
          role={this.props.role}
          name={this.props.name}
          title={this.props.title}
        />
        <div className="Application__main">
          <Switch>
            <Route path="/academy" component={toAcademy(this.props)} />
            <Route path="/playground" component={Playground} />
            <Route path="/login" render={toLogin(this.props)} />
            <Route exact={true} path="/" render={this.redirectToAcademy} />
            <Route component={NotFound} />
          </Switch>
        </div>
      </div>
    );
  }

  private redirectToAcademy = () => <Redirect to="/academy" />;
}

/**
 * A user routes to /academy,
 *  1. If the user is logged in, render the Academy component
 *  2. If the user is not logged in, redirect to /login
 */
const toAcademy = (props: IApplicationProps) =>
  props.accessToken === undefined || props.role === undefined
    ? () => <Redirect to="/login" />
    : () => <Academy accessToken={props.accessToken} role={props.role!} />;

const toLogin = (props: IApplicationProps) => () => (
  <Login ivleToken={qs.parse(props.location.search).token} />
);

const parsePlayground = (props: IApplicationProps) => {
  const prgrm = parsePrgrm(props);
  const chapter = parseChapter(props) || props.currentPlaygroundChapter;
  const externalLibraryName = parseExternalLibrary(props) || props.currentPlaygroundExternalLibrary;
  if (prgrm) {
    props.handleEditorValueChange(prgrm);
    props.handleEnsureLibrariesLoaded();
    props.handleClearContext(chapter, externalLibraryName);
    props.handlePlaygroundExternalSelect(externalLibraryName);
  }
};

const parsePrgrm = (props: RouteComponentProps<{}>) => {
  const qsParsed = qs.parse(props.location.hash);
  // legacy support
  const program = qsParsed.lz !== undefined ? qsParsed.lz : qsParsed.prgrm;
  return program !== undefined ? decompressFromEncodedURIComponent(program) : undefined;
};

const parseChapter = (props: RouteComponentProps<{}>) => {
  const chapQuery = qs.parse(props.location.hash).chap;
  const chap = chapQuery === undefined ? NaN : parseInt(chapQuery, 10);
  return sourceChapters.includes(chap) ? chap : undefined;
};

const parseExternalLibrary = (props: RouteComponentProps<{}>) => {
  const ext = qs.parse(props.location.hash).ext || '';
  return Object.values(ExternalLibraryNames).includes(ext) ? ext : ExternalLibraryNames.NONE;
};

export default Application;
