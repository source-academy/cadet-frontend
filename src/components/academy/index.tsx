import * as React from 'react';
import { Redirect, Route, RouteComponentProps, Switch } from 'react-router';

import Grading from '../../containers/academy/grading';
import AssessmentContainer from '../../containers/assessment';
import Dashboard from '../../containers/dashboard/DashboardContainer';
import StoryUpload from '../../containers/game-dev/StoryUploadContainer';
import Game from '../../containers/GameContainer';
import GroundControl from '../../containers/groundControl/GroundControlContainer';
import MaterialUpload from '../../containers/material/MaterialUploadContainer';
import Sourcereel from '../../containers/sourcecast/SourcereelContainer';
import { isAcademyRe } from '../../reducers/session';
import { Role } from '../../reducers/states';
import { HistoryHelper } from '../../utils/history';
import { assessmentCategoryLink } from '../../utils/paramParseHelpers';
import { AssessmentCategories, AssessmentCategory } from '../assessment/assessmentShape';
import AcademyNavigationBar from './NavigationBar';

interface IAcademyProps extends IOwnProps, IStateProps, IDispatchProps, RouteComponentProps<{}> {}

export interface IOwnProps {
  accessToken?: string;
  role: Role;
}

export interface IStateProps {
  historyHelper: HistoryHelper;
}

export interface IDispatchProps {
  handleFetchNotifications: () => void;
}

const assessmentRenderFactory = (cat: AssessmentCategory) => (
  routerProps: RouteComponentProps<any>
) => <AssessmentContainer assessmentCategory={cat} />;

const assessmentRegExp = ':assessmentId(\\d+)?/:questionId(\\d+)?';
const gradingRegExp = ':submissionId(\\d+)?/:questionId(\\d+)?';

class Academy extends React.Component<IAcademyProps> {
  public componentDidMount() {
    /* TODO: REPLACE WITH LONG POLLING METHOD */
    this.props.handleFetchNotifications();
  }
  public render() {
    return (
      <div className="Academy">
        <AcademyNavigationBar role={this.props.role} />
        <Switch>
          <Route
            path={`/academy/${assessmentCategoryLink(
              AssessmentCategories.Contest
            )}/${assessmentRegExp}`}
            render={assessmentRenderFactory(AssessmentCategories.Contest)}
          />
          <Route path="/academy/game" component={Game} />
          <Route
            path={`/academy/${assessmentCategoryLink(
              AssessmentCategories.Mission
            )}/${assessmentRegExp}`}
            render={assessmentRenderFactory(AssessmentCategories.Mission)}
          />
          <Route
            path={`/academy/${assessmentCategoryLink(
              AssessmentCategories.Path
            )}/${assessmentRegExp}`}
            render={assessmentRenderFactory(AssessmentCategories.Path)}
          />
          <Route
            path={`/academy/${assessmentCategoryLink(
              AssessmentCategories.Sidequest
            )}/${assessmentRegExp}`}
            render={assessmentRenderFactory(AssessmentCategories.Sidequest)}
          />
          <Route
            path={`/academy/${assessmentCategoryLink(
              AssessmentCategories.Practical
            )}/${assessmentRegExp}`}
            render={assessmentRenderFactory(AssessmentCategories.Practical)}
          />
          <Route path="/academy/groundcontrol" component={GroundControl} />
          <Route path="/academy/dashboard" component={Dashboard} />
          <Route path={`/academy/grading/${gradingRegExp}`} component={Grading} />
          <Route path={'/academy/material'} component={MaterialUpload} />
          <Route path="/academy/sourcereel" component={Sourcereel} />
          <Route path={'/academy/gamedev'} component={StoryUpload} />
          <Route exact={true} path="/academy" component={dynamicRedirect(this.props)} />
          <Route component={redirectTo404} />
        </Switch>
      </div>
    );
  }
}

/**
 * 1. If user is in /academy.*, redirect to game
 * 2. If not, redirect to the last /acdaemy.* route the user was in
 * See src/utils/history.ts for more details
 */
const dynamicRedirect = (props: IStateProps) => {
  const clickedFrom = props.historyHelper.lastGeneralLocations[0];
  const lastAcademy = props.historyHelper.lastAcademyLocations[0];
  if (clickedFrom != null && isAcademyRe.exec(clickedFrom!) == null && lastAcademy != null) {
    return () => <Redirect to={lastAcademy!} />;
  } else {
    return redirectToGame;
  }
};

const redirectTo404 = () => <Redirect to="/404" />;

const redirectToGame = () => <Redirect to="/academy/game" />;

export default Academy;
