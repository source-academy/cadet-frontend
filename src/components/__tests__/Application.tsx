import { shallow } from 'enzyme';
import * as React from 'react';

import { mockRouterProps } from '../../mocks/components';
import Application, { IApplicationProps } from '../Application';
import { ExternalLibraryName, ExternalLibraryNames } from '../assessment/assessmentShape';

test('Application renders correctly', () => {
  const props: IApplicationProps = {
    ...mockRouterProps('/academy', {}),
    title: 'Cadet',
    currentPlaygroundChapter: 2,
    handleLogOut: () => {},
    currentPlaygroundExternalLibrary: ExternalLibraryNames.NONE,
    handleClearContext: (chapter: number, externalLibraryName: ExternalLibraryName) => {},
    handleEditorValueChange: (val: string) => {},
    handleEnsureLibrariesLoaded: () => {},
    handlePlaygroundExternalSelect: (externalLibraryName: ExternalLibraryName) => {}
  };
  const app = <Application {...props} />;
  const tree = shallow(app);
  expect(tree.debug()).toMatchSnapshot();
});
