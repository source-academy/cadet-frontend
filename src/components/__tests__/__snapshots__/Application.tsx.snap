// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application renders correctly 1`] = `
"<div className=\\"Application\\">
  <NavigationBar handleLogOut={[Function: handleLogOut]} role={[undefined]} name={[undefined]} title=\\"Cadet\\" />
  <div className=\\"Application__main\\">
    <Switch>
      <Route path=\\"/academy\\" component={[Function]} />
      <Route path=\\"/mission-control/:assessmentId(-?\\\\\\\\d+)?/:questionId(\\\\\\\\d+)?\\" render={[Function: toIncubator]} />
      <Route path=\\"/playground\\" component={[Function: C]} />
      <Route path=\\"/login\\" render={[Function]} />
      <Route path=\\"/contributors\\" component={[function]} />
      <Route path=\\"/sourcecast\\" component={[Function: Connect]} />
      <Route exact={true} path=\\"/\\" render={[Function]} />
      <Route component={[Function: NotFound]} />
    </Switch>
  </div>
</div>"
`;
